lib/spack/spack/test/spec_semantics.py

```diff
@@ -1293,3 +1293,16 @@ def test_package_hash_affects_dunder_and_dag_hash(mock_packages, default_mock_co
     assert hash(a1) != hash(a2)
     assert a1.dag_hash() != a2.dag_hash()
     assert a1.process_hash() != a2.process_hash()
+
+
+def test_satisfies_is_commutative_with_concrete_specs(mock_packages, default_mock_concretization):
+    a1 = default_mock_concretization("a@1.0")
+    a2 = Spec("a@1.0")
+
+    # strict=False means non-empty intersection, which is commutative.
+    assert a1.satisfies(a2)
+    assert a2.satisfies(a1)
+
+    # strict=True means set inclusion, which is not commutative.
+    assert a1.satisfies(a2, strict=True)
+    assert not a2.satisfies(a1, strict=True)
```

Comment:
This doesn't test the initial case we found wrong:
```
Spack version 0.20.0.dev0
Python 3.8.10, Linux x86_64
>>> from spack.spec import Spec
>>> from spack.spec import Spec
>>> s = Spec("zlib")
>>> t = Spec("zlib@1.2.13")
>>> s.satisfies(t)
True
>>> t.satisfies(s)
True
>>> t.concretize()
>>> s.satisfies(t)
False
>>> t.satisfies(s)
True
>>> 
```
where `strict=False` and one spec was concrete.